### PR TITLE
Floating Point

### PR DESCRIPTION
**Pull request by mohammadhossam**
Modified type definitions and type encoders in order to add an initial support in Prusti for floating point numbers.

### CODE REVIEW

**Comment by vakaras** [comment]
Could you [rebase](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) and resolve the conflicts? If you have never done this before, the typical workflow is as follows:

```bash
# Add Prusti main repository.
git remote add upstream git@github.com:viperproject/prusti-dev.git

# Fetch all changes.
git fetch --all

# Start rebase.
git rebase upstream/master

# Follow the instructions on the command line until all conflicts are resolved.

# Update the PR:
git push --force-with-lease
```

**Comment by mohammadhossam** [comment]
I have rebased and resolved conflicts. Please let me know if there are any other needed steps.

File: prusti-viper/src/encoder/encoder.rs [hunk 1]
```diff
@@ -63,10 +63,8 @@ use super::mir::{
         MirTypeEncoderState, MirTypeEncoderInterface,
     },
 };
-use super::high::{
-    pure_functions::{HighPureFunctionEncoderInterface},
-    types::{HighTypeEncoderState, HighTypeEncoderInterface}
-};
+use super::high::types::{HighTypeEncoderState, HighTypeEncoderInterface};
+use vir_crate::polymorphic::FloatConst;
```
Comment (vakaras) [inline prusti-viper/src/encoder/encoder.rs]:
```suggestion

```
It is more consistent with the already existing code to use `vir::FloatConst`.

File: prusti-viper/src/encoder/encoder.rs [hunk 2]
```diff
@@ -715,15 +714,15 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
                 let bits = scalar_value.to_u32().unwrap();
                 vir::Expr::Const(
                     vir::ConstExpr {
-                        value: vir::Const::Float(vir::FloatConst::F32(bits)),
+                        value: vir::Const::Float(FloatConst::F32(bits)),
```
Comment (vakaras) [inline prusti-viper/src/encoder/encoder.rs]: This should be (also in other places):
```suggestion
                        value: vir::Const::Float(vir::FloatConst::F32(bits)),
```

File: prusti-viper/src/encoder/high/lower/expression.rs [hunk 3]
```diff
@@ -1,7 +1,8 @@
 use super::super::types::interface::HighTypeEncoderInterfacePrivate;
 
 use super::IntoPolymorphic;
-use vir_crate::{high as vir_high, polymorphic as vir_poly};
+use vir_crate::{high as vir_high, legacy::FloatConst, polymorphic as vir_poly};
+// use vir_crate::polymorphic::FloatConst;
```
Comment (vakaras) [inline prusti-viper/src/encoder/high/lower/expression.rs]:
```suggestion

```

File: prusti-viper/src/encoder/high/lower/expression.rs [hunk 4]
```diff
@@ -1,7 +1,8 @@
 use super::super::types::interface::HighTypeEncoderInterfacePrivate;
 
 use super::IntoPolymorphic;
-use vir_crate::{high as vir_high, polymorphic as vir_poly};
+use vir_crate::{high as vir_high, legacy::FloatConst, polymorphic as vir_poly};
```
Comment (vakaras) [inline prusti-viper/src/encoder/high/lower/expression.rs]:
```suggestion
use vir_crate::{high as vir_high, legacy as vir_low, polymorphic as vir_poly};
```
and use `vir_low::FloatConst`

**Review by vakaras — APPROVED** [review]
I have added some final nits. I think that once they are fixed, the PR can be merged.

**Comment by mohammadhossam** [comment]
Done! I have fixed them.

**Review by vakaras — APPROVED** [review]
Looks great!

**Pull request merged**